src/main/java/org/folio/circulation/domain/validation/ItemLimitValidator.java

```diff
@@ -0,0 +1,164 @@
+
+  public ItemLimitValidator(Function<String, ValidationErrorFailure> itemLimitErrorFunction,
+    LoanRepository loanRepository, LoanPolicyRepository loanPolicyRepository,
+    CollectionResourceClient circulationRulesStorage, RoutingContext routingContext) {
+
+    this.itemLimitErrorFunction = itemLimitErrorFunction;
+    this.loanRepository = loanRepository;
+    this.loanPolicyRepository = loanPolicyRepository;
+    this.circulationRulesStorage = circulationRulesStorage;
+    this.routingContext = routingContext;
+  }
+
+  public CompletableFuture<Result<LoanAndRelatedRecords>> refuseWhenItemLimitIsReached(
+    LoanAndRelatedRecords records) {
+
+    Loan loan = records.getLoan();
+    Integer itemLimit = loan.getLoanPolicy().getItemLimit();
+
+    if (itemLimit == null) {
+      return completedFuture(succeeded(records));
+    }
+
+    return loanPolicyRepository.lookupLineNumber(loan.getItem(), loan.getUser())
+      .thenComposeAsync(r -> r.after(this::retrieveCirculationRule)
+        .thenComposeAsync(result -> result.failAfter(rule -> isLimitReached(rule, records),
+          rule -> {
+            String message = getErrorMessage(rule);
+            return itemLimitErrorFunction.apply(String.format("Patron has reached maximum item limit of %d items %s",
+              itemLimit, message));
+          }))
+        .thenApply(result -> result.map(v -> records)));
+  }
+
+  private CompletableFuture<Result<Boolean>> isLimitReached(String rule, LoanAndRelatedRecords records) {
+    Item item = records.getLoan().getItem();
+    String materialType = item.getMaterialType().getString("name");
+    String loanType = item.getLoanTypeName();
+    Integer itemLimit = records.getLoan().getLoanPolicy().getItemLimit();
+
+    if (isRuleMaterialTypePresent(rule) || isRuleLoanTypePresent(rule)) {
+      return loanRepository.findOpenLoansByUserIdAndLoanPolicyIdWithItem(records)
+        .thenApply(r -> r.map(loans -> loans.getRecords().stream()
+            .filter(loan -> isMaterialTypeMatchInRetrievedLoan(rule, materialType, loan))
+            .filter(loan -> isLoanTypeMatchInRetrievedLoan(rule, loanType, loan))
+            .count()))
+        .thenApply(r -> r.map(loansCount ->loansCount >= itemLimit));
+    }
+
+    return loanRepository.findOpenLoansByUserIdAndLoanPolicyId(records)
+      .thenApply(r -> r.map(loans -> loans.getTotalRecords() >= itemLimit));
+  }
+
+  private boolean isMaterialTypeMatchInRetrievedLoan(String rule, String expectedMaterialType, Loan loan) {
+    if (isRuleMaterialTypePresent(rule)) {
+      if (expectedMaterialType.equals(loan.getItem().getMaterialType().getString("name"))) {
+        return true;
+      } else {
+        return false;
+      }
+    } else {
+      return true;
+    }
+  }
+
+  private boolean isLoanTypeMatchInRetrievedLoan(String rule, String expectedLoanType, Loan loan) {
+    if (isRuleLoanTypePresent(rule)) {
+      if (expectedLoanType.equals(loan.getItem().getLoanTypeName())) {
```

Comment:
Could this be changed to
`return expectedMaterialType.equals(loan.getItem().getMaterialType().getString("name")`

Reply:
thanks, updated